src/NServiceBus.AzureStoragePersistence/SagaPersisters/Azure/SecondaryIndeces/SecondaryIndexPersister.cs

```diff
@@ -0,0 +1,215 @@
+            if (sagaId == null)
+            {
+                return null;
+            }
+
+            var entity = new SecondaryIndexTableEntity();
+            key.Apply(entity);
+            entity.SagaId = sagaId.Value;
+
+            try
+            {
+                table.Execute(TableOperation.Insert(entity));
+            }
+            catch (StorageException)
+            {
+                throw new RetryNeededException();
+            }
+
+            cache.Put(key, sagaId.Value);
+            return sagaId;
+        }
+
+        private static bool IsConflict(StorageException ex)
+        {
+            return ex.RequestInformation.HttpStatusCode == (int) HttpStatusCode.Conflict;
+        }
+
+        private class IndexDefintion
+        {
+            static readonly ConcurrentDictionary<Type, object> sagaToIndex = new ConcurrentDictionary<Type, object>();
+            static readonly object NullValue = new object();
+
+            static readonly ParameterExpression ObjectParameter = Expression.Parameter(typeof(object));
+
+            readonly string sagaTypeName;
+            readonly string propertyName;
+
+            public Func<object, object> Accessor { get; }
+
+            private IndexDefintion(Type sagaType, PropertyInfo pi)
+            {
+                sagaTypeName = sagaType.FullName;
+                propertyName = pi.Name;
+                Accessor = Expression.Lambda<Func<object, object>>(Expression.Convert(Expression.MakeMemberAccess(Expression.Convert(ObjectParameter, sagaType), pi), typeof(object)), ObjectParameter).Compile();
+            }
+
+            public static IndexDefintion Get(Type sagaType)
+            {
+                var index = sagaToIndex.GetOrAdd(sagaType, type =>
+                {
+                    var pi = UniqueAttribute.GetUniqueProperty(sagaType);
+                    if (pi == null)
+                    {
+                        return NullValue;
+                    }
+
+                    return new IndexDefintion(sagaType, pi);
+                });
+
+                if (ReferenceEquals(index, NullValue))
+                {
+                    return null;
+                }
+
+                return (IndexDefintion) index;
+            }
+
+            public void ValidateProperty(string propertyName)
+            {
+                if (this.propertyName != propertyName)
+                {
+                    throw new ArgumentException($"The following saga '{sagaTypeName} is not indexed by '{propertyName}'. The only secondary index is defined for '{this.propertyName}'. " +
+                                                $"Ensure that the saga is correlated properly.");
+                }
+            }
+
+            public PartitionRowKeyTuple BuildTableKey(object propertyValue)
+            {
+                return new PartitionRowKeyTuple($"Index_{sagaTypeName}", $"{propertyName}_{Serialize(propertyValue)}");
```

Comment:
@yvesgoeleven @SeanFeldman I've just read this and I'm not sure about the performance, I need your knowledge about WAS internals. This line above results in having all the indexes of the given saga type under one partition key. Should I use the property value as well for it, just to enable more partitions? AFAIK in the scalability targets for WAS there is no partition limitations (they are only for an account).